src/Access/systemDBGrantsToElements.cpp

```diff
@@ -0,0 +1,102 @@
+#include <Access/systemDBGrantsToElements.h>
+#include <Access/Common/SystemDBGrants.h>
+#include <Interpreters/DatabaseCatalog.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    AccessRightsElements getSystemDBGrantsElements_All()
+    {
+        AccessRightsElements elements;
+        AccessFlags select_flags = AccessType::SELECT;
+        elements.emplace_back(select_flags, DatabaseCatalog::SYSTEM_DATABASE);
+        return elements;
+    }
+
+    AccessRightsElements getImplicitSystemDBGrantsImpl()
+    {
+        AccessRightsElements elements;
+
+        /// Constant tables.
+        static const char * tables_in_system_db[] = {
+            "one",
+            "numbers",
+            "numbers_mt",
+            "zeros",
+            "zeros_mt",
+            "functions",
+            "build_options",
+            "formats",
+            "table_functions",
+            "aggregate_function_combinators",
+            "data_type_families",
+            "collations",
+            "table_engines",
+            "contributors",
+            "privileges",
+            "licenses",
+            "time_zones"};
+        AccessFlags select_flags = AccessType::SELECT;
+        for (const auto * table : tables_in_system_db)
+            elements.emplace_back(select_flags, DatabaseCatalog::SYSTEM_DATABASE, table);
+
+        /// Constant columns of non-constant tables.
+        elements.emplace_back(select_flags, DatabaseCatalog::SYSTEM_DATABASE, "settings", std::vector<std::string_view>{"name", "description", "type"});
+        elements.emplace_back(select_flags, DatabaseCatalog::SYSTEM_DATABASE, "merge_tree_settings", std::vector<std::string_view>{"name", "description", "type"});
+        elements.emplace_back(select_flags, DatabaseCatalog::SYSTEM_DATABASE, "replicated_merge_tree_settings", std::vector<std::string_view>{"name", "description", "type"});
+
+        /// We implicitly grant access to tables in the `information_schema` database because those tables are implemented
+        /// via requests to some tables in the `system` database (system.databases -> information_schema.schemata,
+        /// system.tables -> information_schema.tables, system.tables -> information_schema.views,
+        /// system.columns -> information_schema.columns).
+        /// So for example to perform SELECT from `information_schema.schemata` it's necessary to have an ability to SELECT from
+        /// system.databases.
+        static const char * tables_in_information_schema_db[] = {"schemata", "tables", "views", "columns"};
+        for (const auto * table : tables_in_information_schema_db)
+            elements.emplace_back(select_flags, DatabaseCatalog::INFORMATION_SCHEMA, table);
+
+        static const char * tables_in_information_schema_db_uppercased[] = {"SCHEMATA", "TABLES", "VIEWS", "COLUMNS"};
```

Comment:
This can be written with [IIFE](https://www.cppstories.com/2016/11/iife-for-complex-initialization/). And because the variable is static it will be called only once.
Because the names are connected to the list above, it would be bad if we added `information_schema` table to only one of the arrays.